### PR TITLE
WIP Super-resolution tool - phase 1

### PR DESCRIPTION
**Pull request by mafonsoN**
## Description

First PR for the implementation of the frame super-resolution tool in svt-av1. Super-resolution is implemented for intra-only frames. Should be working for all enc-modes and also 10-bit inputs.

New command line arguments added: 
`-superres-mode` (0: None, 1: Fixed denominator, 2: Random denominator)
`-superres-denom` (denominator, d, used to determine encoding width = (8 * width + d/2 ) / d
`-superres-kf-denom` (denominator for key frames)

Example command line for super-res with x2 downsampling (d=16)
`./SvtAv1EncApp -i input.y4m -b out_bitstream.ivf -bit-depth 8 -enc-mode 6 -stat-report 1 -stat-file stat.txt -superres-mode 1 -superres-kf-denom 16`

Limitations (to be resolved before merging):

- [ ] Small rec/dec mismatch for some SBs
- [ ] Downsampled width needs to be multiple of 8 or the stream is corrupted
- [ ] No unit tests yet

For the next phase:

- Support for inter frame scaling
- 2-pass support
- More superres-modes

## Performance

No expected bitstream impact when super-res is disabled (default).

Performance impact when enabled TBD.

### CODE REVIEW

File: Source/App/EncApp/EbAppConfig.c [hunk 1]
```diff
@@ -109,6 +109,12 @@
 #define ALTREF_NFRAMES "-altref-nframes"
 #define ENABLE_OVERLAYS "-enable-overlays"
 // --- end: ALTREF_FILTERING_SUPPORT
+// --- start: ALTREF_FILTERING_SUPPORT
```
Comment (kylophone) [inline Source/App/EncApp/EbAppConfig.c]: typo

File: Source/Lib/Encoder/Codec/EbPictureControlSet.c [hunk 2]
```diff
@@ -1277,9 +1298,192 @@ EbErrorType picture_parent_control_set_ctor(PictureParentControlSet *object_ptr,
...
+
+EbErrorType derive_input_resolution_pcs(PictureParentControlSet *pcs_ptr, uint32_t inputSize) {
+    EbErrorType return_error = EB_ErrorNone;
+
+    pcs_ptr->input_resolution =
+            (inputSize < INPUT_SIZE_1080i_TH)
+            ? INPUT_SIZE_576p_RANGE_OR_LOWER
+            : (inputSize < INPUT_SIZE_1080p_TH)
+              ? INPUT_SIZE_1080i_RANGE
+              : (inputSize < INPUT_SIZE_4K_TH) ? INPUT_SIZE_1080p_RANGE : INPUT_SIZE_4K_RANGE;
```
Comment (kylophone) [inline Source/Lib/Encoder/Codec/EbPictureControlSet.c]: There might be a nicer way to write this. ternary into a ternary into a ternary.

File: Source/App/EncApp/EbAppConfig.c (hunk 1 above)
Comment (mafonsoN) [reply Source/App/EncApp/EbAppConfig.c]: thanks @kylophone, fixed.

File: Source/Lib/Encoder/Codec/EbPictureControlSet.c (hunk 2 above)
Comment (mafonsoN) [reply Source/Lib/Encoder/Codec/EbPictureControlSet.c]: Addressed.

File: Source/App/EncApp/EbAppConfig.c [hunk 3]
```diff
@@ -473,6 +479,20 @@ static void set_enable_overlays(const char *value, EbConfig *cfg) {
     cfg->enable_overlays = (EbBool)strtoul(value, NULL, 0);
 };
 // --- end: ALTREF_FILTERING_SUPPORT
+// --- start: SUPER-RESOLUTION SUPPORT
+static void SetSuperresMode(const char *value, EbConfig *cfg) {
```
Comment (hassount) [inline Source/App/EncApp/EbAppConfig.c]: snake_case for function names

File: Source/Lib/Encoder/Codec/EbEncDecProcess.c [hunk 4]
```diff
@@ -665,7 +667,7 @@ void psnr_calculations(PictureControlSet *pcs_ptr, SequenceControlSet *scs_ptr)
             recon_ptr = pcs_ptr->recon_picture_ptr;
 
         EbPictureBufferDesc *input_picture_ptr =
-            (EbPictureBufferDesc *)pcs_ptr->parent_pcs_ptr->enhanced_picture_ptr;
+            (EbPictureBufferDesc *)pcs_ptr->parent_pcs_ptr->enhanced_unscaled_picture_ptr;
```
Comment (hassount) [inline Source/Lib/Encoder/Codec/EbEncDecProcess.c]: maybe nonscaled is more appropriate as a name

File: Source/Lib/Encoder/Codec/EbModeDecisionConfiguration.c [hunk 5]
```diff
@@ -248,13 +248,12 @@ EbErrorType mdc_refinement(
...
 
 void refinement_prediction_loop(
-    SequenceControlSet                   *scs_ptr,
     PictureControlSet                    *pcs_ptr,
     uint32_t                              sb_index,
     ModeDecisionConfigurationContext     *context_ptr)
 {
     MdcpLocalBlkStruct    *local_blk_array         = context_ptr->local_blk_array;
-    SbParams               *sb_params            = &scs_ptr->sb_params_array[sb_index];
+    SbParams               *sb_params            = &pcs_ptr->parent_pcs_ptr->sb_params_array[sb_index];
```
Comment (hassount) [inline Source/Lib/Encoder/Codec/EbModeDecisionConfiguration.c]: this might go beyond the col limits for clang, was clang-format applied on this code ?

File: Source/Lib/Encoder/Codec/EbResize.c [hunk 6]
```diff
@@ -0,0 +1,1020 @@
+/*
```
Comment (hassount) [inline Source/Lib/Encoder/Codec/EbResize.c]: was this file brought as a whole from libaom ? if so, might b best to have the name stay the same probably resize.c

File: Source/Lib/Encoder/Codec/EbResize.h [hunk 7]
```diff
@@ -0,0 +1,216 @@
+#ifndef SVT_AV1_SCALE_H
```
Comment (hassount) [inline Source/Lib/Encoder/Codec/EbResize.h]: missing license / copyright header

File: Source/Lib/Encoder/Codec/EbRestProcess.c [hunk 8]
```diff
@@ -230,6 +244,209 @@ void get_own_recon(SequenceControlSet *scs_ptr, PictureControlSet *pcs_ptr,
...
+
+    return EB_ErrorNone;
+}
+
+void get_recon_pic(PictureControlSet *pcs_ptr,
+                   EbPictureBufferDesc **recon_ptr,
+                   EbBool is_highbd){
+    if(!is_highbd){
+        if (pcs_ptr->parent_pcs_ptr->is_used_as_reference_flag == EB_TRUE)
+            *recon_ptr = ((EbReferenceObject*)pcs_ptr->parent_pcs_ptr->reference_picture_wrapper_ptr->object_ptr)->reference_picture;
```
Comment (hassount) [inline Source/Lib/Encoder/Codec/EbRestProcess.c]: clang-format ?

File: Source/Lib/Encoder/Codec/EbModeDecisionConfiguration.c (hunk 5 above)
Comment (mafonsoN) [reply Source/Lib/Encoder/Codec/EbModeDecisionConfiguration.c]: No I haven't applied. Will apply it on the diff. Should we be testing it automatically as well?

File: Source/Lib/Encoder/Codec/EbResize.c (hunk 6 above)
Comment (mafonsoN) [reply Source/Lib/Encoder/Codec/EbResize.c]: No, this is not all copied from libaom. I'll add the license. Thanks

File: Source/Lib/Encoder/Codec/EbModeDecisionConfiguration.c (hunk 5 above)
Comment (hassount) [reply Source/Lib/Encoder/Codec/EbModeDecisionConfiguration.c]: yes, part of the todo list for it to be applied automatically on the PRs, that said, the way its set right now, it applies automatically for the IDE once you save, but such depends on the environment.

File: Source/Lib/Encoder/Codec/EbEncDecProcess.c (hunk 4 above)
Comment (mafonsoN) [reply Source/Lib/Encoder/Codec/EbEncDecProcess.c]: @hassount unscaled is used in libaom and in the decoder. Do you have a strong preference to nonscaled?

File: Source/Lib/Encoder/Codec/EbResize.h [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+ *
+ * This source code is subject to the terms of the BSD 2 Clause License and
+ * the Alliance for Open Media Patent License 1.0. If the BSD 2 Clause License
+ * was not distributed with this source code in the LICENSE file, you can
+ * obtain it at www.aomedia.org/license/software. If the Alliance for Open
+ * Media Patent License 1.0 was not distributed with this source code in the
+ * PATENTS file, you can obtain it at www.aomedia.org/license/patent.
+ */
+/*
+* Copyright(c) 2019 Intel Corporation
```
Comment (chkngit) [inline Source/Lib/Encoder/Codec/EbResize.h]: we have to do resizing for HME, is there a way to unify/take advantge of that?

File: Source/Lib/Encoder/Codec/EbRestProcess.c [hunk 10]
```diff
@@ -267,11 +486,26 @@ void *rest_kernel(void *input_ptr) {
...
+
+            // ------- start: Normative upscaling - super-resolution tool
+            if(!av1_superres_unscaled(&cm->frm_size)) {
+
+                eb_av1_superres_upscale_frame(cm,
+                                              pcs_ptr,
+                                              scs_ptr);
+
+                uint16_t picture_sb_width = (uint16_t)(
+                        (cm->frm_size.superres_upscaled_width + scs_ptr->sb_sz - 1) / scs_ptr->sb_sz);
```
Comment (chkngit) [inline Source/Lib/Encoder/Codec/EbRestProcess.c]: does this support the restoration segments (parallelizion tool)

File: Source/Lib/Encoder/Codec/EbRestoration.c [hunk 11]
```diff
@@ -1590,10 +1590,12 @@ int32_t eb_av1_loop_restoration_corners_in_sb(Av1Common *cm, int32_t plane, int3
     //
     // from which we get u = D * MI_SIZE * m / N
 
-    const int32_t mi_to_num_x =
-        mi_size_x; //CHKN  av1_superres_unscaled(cm) ? mi_size_x : mi_size_x * cm->superres_scale_denominator;
-    const int32_t mi_to_num_y = mi_size_y;
-    const int32_t denom_x = size; //CHKN  av1_superres_unscaled(cm) ? size : size * SCALE_NUMERATOR;
+    // M'notes: Changes need to be made here - probably the same as libaom
+    const int mi_to_num_x = !av1_superres_unscaled(&cm->frm_size)
```
Comment (chkngit) [inline Source/Lib/Encoder/Codec/EbRestoration.c]: I hope it was easy for you to figure that you need to enable those changes here, kept them commented out for this purpose!!

File: Source/Lib/Encoder/Codec/EbSequenceControlSet.h [hunk 12]
```diff
@@ -232,7 +232,7 @@ extern EbErrorType eb_sequence_control_set_instance_ctor(
 
 extern EbErrorType sb_params_init(SequenceControlSet *scs_ptr);
 
-extern EbErrorType derive_input_resolution(SequenceControlSet *scs_ptr, uint32_t input_size);
```
Comment (chkngit) [inline Source/Lib/Encoder/Codec/EbSequenceControlSet.h]: can you also remove the old structures(gem, and width/height) from Seq so that people will not continue to use them, and use the picture instead?

File: Source/Lib/Encoder/Codec/EbRestoration.c (hunk 11 above)
Comment (mafonsoN) [reply Source/Lib/Encoder/Codec/EbRestoration.c]: It did, thanks.

File: Source/Lib/Encoder/Codec/EbPictureControlSet.h [hunk 13]
```diff
@@ -940,6 +944,21 @@ typedef struct PictureParentControlSet {
 #endif
     uint8_t tx_size_early_exit;
 
+    SbParams *sb_params_array;
+    SbGeom *sb_geom;
```
[the remark this answers is not in the document]
Comment (mafonsoN) [reply Source/Lib/Encoder/Codec/EbPictureControlSet.h]: There are 9 possible resolutions actually (only widths are changed for now). It would be a waste to allocate all possible ones on the sequence level. Right now it only allocates the input resolution on a sequence level and sets these picture pointers to the sequence level ones. Then, in the case that the resolution of any frame is changed, it will allocate another one and that pointer will point to the picture one instead. Does this make sense? Anyway, I'll review to check if I am over allocating.

File: Source/Lib/Encoder/Codec/EbResize.h (hunk 9 above)
Comment (mafonsoN) [reply Source/Lib/Encoder/Codec/EbResize.h]: We might, but it's better to hold this thought for the next PR, where inter pictures will be able to be resized as well.

File: Source/Lib/Encoder/Codec/EbSequenceControlSet.h (hunk 12 above)
Comment (mafonsoN) [reply Source/Lib/Encoder/Codec/EbSequenceControlSet.h]: Yes, this is a good point.

File: Source/Lib/Encoder/Codec/EbPictureControlSet.h [hunk 14]
```diff
@@ -940,6 +944,21 @@ typedef struct PictureParentControlSet {
...
+    EbInputResolution input_resolution;
+    uint16_t picture_sb_width;
+    uint16_t picture_sb_height;
+    uint16_t sb_total_count_unscaled;
+
+    uint16_t aligned_width;
```
[the remark this answers is not in the document]
Comment (mafonsoN) [reply Source/Lib/Encoder/Codec/EbPictureControlSet.h]: I agree, will do.

File: Source/Lib/Encoder/Codec/EbPictureControlSet.h [hunk 15]
```diff
@@ -610,6 +612,8 @@ typedef struct PictureParentControlSet {
     EbObjectWrapper *    reference_picture_wrapper_ptr;
     EbObjectWrapper *    pa_reference_picture_wrapper_ptr;
     EbPictureBufferDesc *enhanced_picture_ptr;
+    EbPictureBufferDesc *enhanced_downscaled_picture_ptr;
+    EbPictureBufferDesc *enhanced_unscaled_picture_ptr;
```
[the remark this answers is not in the document]
Comment (mafonsoN) [reply Source/Lib/Encoder/Codec/EbPictureControlSet.h]: This was because the scaled one is used much more often the non-scaled one, so less code changes need to be done. Does this make sense?

File: Source/Lib/Encoder/Codec/EbPictureControlSet.h [hunk 16]
```diff
@@ -551,6 +551,8 @@ typedef struct PictureControlSet {
     uint16_t tile_row_count;
     uint16_t tile_column_count;
 #endif
+    uint16_t sb_total_count_pix;
+    uint16_t sb_total_count_unscaled;
```
[the remark this answers is not in the document]
Comment (mafonsoN) [reply Source/Lib/Encoder/Codec/EbPictureControlSet.h]: the term pix was used to refer to the total number of sbs of size scs_ptr->sb_size_pix which is different from scs_ptr->sb_sz (one goes up to 128 and the other to 64, if I am not wrong). But I agree it is confusing, I'll re-think this.

**Comment by OmranAbdelkafi** [comment]
ran some speed and bd-rate checks for encoder mode 2 and encoder mode 0 : 
confirmed lossless changes with no performance impact

**Comment by mafonsoN** [comment]
> ran some speed and bd-rate checks for encoder mode 2 and encoder mode 0 :
> confirmed lossless changes with no performance impact

Thanks @OmranAbdelkafi for running these tests

File: Source/Lib/Encoder/Codec/EbInitialRateControlProcess.c [hunk 17]
```diff
@@ -869,6 +870,16 @@ void *initial_rate_control_kernel(void *input_ptr) {
...
             release_pa_reference_objects(scs_ptr, pcs_ptr);
 
+            //****************************************************
+            // Picture resizing for super-res tool
+            //****************************************************
+
+            // Scale picture if super-res is used
+            if(scs_ptr->static_config.superres_mode > SUPERRES_NONE){
+                init_resize_picture(pcs_ptr->scs_ptr,
+                                    pcs_ptr);
```
[the remark this answers is not in the document]
Comment (mafonsoN) [reply Source/Lib/Encoder/Codec/EbInitialRateControlProcess.c]: This will be moved to another process in the next PR.

File: Source/Lib/Encoder/Codec/EbPictureControlSet.h (hunk 14 above)
[the remark this answers is not in the document]
Comment (mafonsoN) [reply Source/Lib/Encoder/Codec/EbPictureControlSet.h]: Done

File: Source/Lib/Encoder/Codec/EbPictureManagerProcess.c [hunk 18]
```diff
@@ -956,11 +984,71 @@ void *picture_manager_kernel(void *input_ptr) {
...
+
+                        if(entry_pcs_ptr->frame_superres_enabled){
+                            // Modify sb_prt_array in child pcs
+                            uint16_t    sb_index;
+                            uint16_t    sb_origin_x = 0;
+                            uint16_t    sb_origin_y = 0;
+                            for (sb_index = 0; sb_index < child_pcs_ptr->sb_total_count_pix; ++sb_index) {
+                                largest_coding_unit_dctor(child_pcs_ptr->sb_ptr_array[sb_index]);
+                                largest_coding_unit_ctor(child_pcs_ptr->sb_ptr_array[sb_index],
+                                                         (uint8_t)scs_ptr->sb_size_pix,
```
[the remark this answers is not in the document]
Comment (mafonsoN) [reply Source/Lib/Encoder/Codec/EbPictureManagerProcess.c]: This is not called when the tool is off. There are a lot of aspects to be considered here. The plan is move this in the next PR.

File: Source/Lib/Encoder/Codec/EbSequenceControlSet.h (hunk 12 above)
Comment (mafonsoN) [reply Source/Lib/Encoder/Codec/EbSequenceControlSet.h]: This needs to be thought about further in order to avoid allocating too much memory. All frames encoded at the source resolution can share these structures. This will be addressed in the next PR.

File: Source/Lib/Encoder/Codec/EbRestProcess.c (hunk 10 above)
Comment (mafonsoN) [reply Source/Lib/Encoder/Codec/EbRestProcess.c]: I believe so, everything else stays the same.

File: Source/Lib/Encoder/Codec/EbPictureManagerProcess.c [hunk 19]
```diff
@@ -819,22 +841,28 @@ void *picture_manager_kernel(void *input_ptr) {
...
+
 #if TILES_PARALLEL
+                        set_tile_info(entry_pcs_ptr);
+
                         int      sb_size_log2    = entry_scs_ptr->seq_header.sb_size_log2;
-                        uint32_t encDecSegColCnt = entry_scs_ptr->enc_dec_segment_col_count_array
-                                                       [entry_pcs_ptr->temporal_layer_index];
+                        uint32_t encDecSegColCnt = (scs_ptr->static_config.super_block_size == 128) ?
+                                                 ((entry_pcs_ptr->aligned_width + 64) / 128) :
+                                                 ((entry_pcs_ptr->aligned_width + 32) / 64);
```
Comment (hassount) [inline Source/Lib/Encoder/Codec/EbPictureManagerProcess.c]: @mafonsoN was this change intentional ?

Comment (hassount) [reply Source/Lib/Encoder/Codec/EbPictureManagerProcess.c]: @andrey-norkin FYI to fix
